### DISCUSSION TITLE
@validates with multiple attributes

### DISCUSSION THREAD

**Discussion by karmingc** [sqlalchemy/sqlalchemy#7947] [Usage Questions] [answered]
Is it possible to validate multiple attributes with a single @validates decorator?

For example, assuming my model `Person` has `first_name`, `last_name`, is there way to validate both attributes?

With one attribute 
```python

Person(first_name="test_f", last_name="test_l")

@validates("first_name")
def validate_ex(self, key, value)
    print(value)
>>> "test_f"
  
```

With two attribute

```python
Person(first_name="test_f", last_name="test_l")

@validates("first_name", "last_name")
def validate_ex(self, key, value, last_name)
    print(value)
    print(last_name)
>>> TypeError: .... () missing 1 required positional argument: 'last_name'
```

Alternatively, would there be a more appropriate approach to validating multiple attributes?

**Comment by karmingc** [comment]
sorry for the ping @zzzeek, but would you have any suggestions?

**Comment by zzzeek** [comment, marked as answer]
yes validates accepts any number of names, passing each name as encountered to the validation function , see https://docs.sqlalchemy.org/en/14/orm/mapped_attributes.html?highlight=validates#sqlalchemy.orm.validates

**Reply by karmingc** [reply]
Yes, i can see from the documentations as well that it accepts `names` and `kw`. However, i'm not able to reproduce the functionality. From the code I provided, is there anything I'm doing wrong?

**Reply by zzzeek** [reply]
yes,the validator takes exactly the "key" and "value" arguments:

```py
@validates("first_name", "last_name")
def validate_ex(self, key, value)
    print(f"The value of the attriribute named {key} is {value}")
```

**Reply by karmingc** [reply]
Yes, so that would be the first attribute, but how should one proceed to access the second attribute, which is `last_name` in our example?

**Reply by karmingc** [reply]
Oh, from my tests, it seems that it's not evaluating both simultaneously, but rather iterating through?

**Reply by karmingc** [reply]
A slight update, I decided to leverage [instance events](https://docs.sqlalchemy.org/en/14/orm/events.html#instance-events) and validate the keyword arguments at that point since @validates doesn't seem to support simultaneous checks. Any thoughts?

**Reply by zzzeek** [reply]
if you are seeking to validate within instantiation, that is, the `__init__` method, the most straightforward approach is to add the validation code directly into the `__init__` method.     

however this won't accommodate for the case where someone updates the values on an existing object.  for that approach, the most traditional would be a method such as `set_fullname()`.   However you could also seek to represent the multiple fields using a [composite attribute](https://docs.sqlalchemy.org/en/14/orm/composites.html?highlight=composite#mapper-composite), meaning the value would be set using a new kind of object type that spans multiple columns, which would allow `@validates` to get the fully composed value at once.  

finally, the ultimate way to validate everything about an object is to use persistence events like [before_insert](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_insert#sqlalchemy.orm.MapperEvents.before_insert) and [before_update](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_update#sqlalchemy.orm.MapperEvents.before_update), but these would only affect change at the point at which the object's state is persisted.

my preference would be traditional use of `__init__` and `set_fullname()`.

**Reply by karmingc** [reply]
Thank you for all your suggestions, I appreciate it!

Concerning the `set_fullname()`, would you happen to have an example? I'm not sure I have an idea how that works..

**Reply by zzzeek** [reply]
```py

class MyClass(Base):
    __tablename__ = 'my_table'

    # ...

    first_name = Column(...)
    last_name = Column(...)

    def __init__(self, first_name, last_name, **kw):
        self.set_fullname(first_name, last_name)
        super().__init__(**kw)

    def set_fullname(self, first_name, last_name):
        if not first_name or not last_name:
            raise ValueError("first and last name are required")
        self.first_name = first_name
        self.last_name = last_name
```

**Reply by karmingc** [reply]
lovely, thank you @zzzeek for all your help!
